solarforecastarbiter/metrics/calculator.py

```diff
@@ -3,98 +3,161 @@
-            rmse[idx] = deterministic.root_mean_square(
-                group.obs_values, group.fx_values)
-        metrics["day"]["mae"] = pd.Series(mae)
-        metrics["day"]["mbe"] = pd.Series(mbe)
-        metrics["day"]["rmse"] = pd.Series(rmse)
-
-        # hourly metrics
-        mae, mbe, rmse = {}, {}, {}
-        for idx, group in df.groupby(df.index.hour):
-            mae[idx] = deterministic.mean_absolute(
-                group.obs_values, group.fx_values)
-            mbe[idx] = deterministic.mean_bias(
-                group.obs_values, group.fx_values)
-            rmse[idx] = deterministic.root_mean_square(
-                group.obs_values, group.fx_values)
-        metrics["hour"]["mae"] = pd.Series(mae)
-        metrics["hour"]["mbe"] = pd.Series(mbe)
-        metrics["hour"]["rmse"] = pd.Series(rmse)
-
-    return metrics
+        return metric_func(obs, fx)
+
+
+def calculate_deterministic_metrics(processed_fx_obs, categories, metrics,
+                                    ref_fx_obs=None, normalizer=1.0):
+    """
+    Calculate deterministic metrics for the processed data using the provided
+    categories and metric types.
+
+    Parameters
+    ----------
+    processed_fx_obs :
+        solarforecastarbiter.datamodel.ProcessedForecastObservation
+    categories : list of str
+        List of categories to compute metrics over.
+    metrics : list of str
+        List of metrics to be computed.
+    ref_fx_obs :
+        solarforecastarbiter.datamodel.ProcessedForecastObservation
+        Reference forecast to be used when calculating skill metrics. Default
+        is None and no skill metrics will be calculated.
+    normalizer : float
+        Normalized factor (should be in the same units as data). Default is
+        1.0 and only needed for normalized metrics.
+
+    Returns
+    -------
+    pd.DataFrame or dict:
+        Contains all the computed metrics by categories.
+        Structure is:
+        1. Category type as tuple (e.g., ('total'), ('month', 'hour'))
+        2. Metric name (e.g., 'mae', 'rmse')
+        3. Category group (e.g, 0, 1, 2 ..., 11 for month)
+        4. Value
+        If no forecast data is found an empty dictionary is returned.
+    """
+    calc_metrics = defaultdict(dict)
+    fx = processed_fx_obs.forecast_values
+    obs = processed_fx_obs.observation_values
+
+    # Check reference forecast is from processed pair, if needed
+    ref_fx = None
+    if any(m in deterministic._REQ_REF_FX for m in metrics):
+        ref_fx = ref_fx_obs.forecast_values
+
+    # No data or metrics
+    if fx.empty or obs.empty or len(metrics) == 0:
+        return calc_metrics
+
+    # Calculate metrics
+    for category in set(categories):
+        calc_metrics[category] = {}
+
+        # total (special category)
+        if category == 'total':
+            for metric_ in metrics:
+                r = _apply_deterministic_metric_func(metric_, fx, obs,
```

Comment:
is `r` for "result"? a little confusing in this context since `r` is also a metric. `out` or `res` or similar?

Reply:
Updating